{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## RDP Classifier\n",
    "\n",
    "Three datastes are required which are train, validation, and test. Each dataset contains sequence ID, sequence, and taxonomy ranks from kingdom to species. Train and validation dataset will be merged into one, becuase machine learning does not require validation stpes in training. Databases are converted to ready4train taxonomy and fasta files to be utilized by RDP Classifier.\n",
    "\n",
    "RDP Classifier need to be downloaded to make models and classify. It can be downloaded from: [RDP Classifier](https://github.com/rdpstaff/classifier). \n",
    "\n",
    "After classification, output text file shows the best predictions for each sequences with confidence values. Confidence value can be changed. Default is 0.8. Classificatoin output is compared with true value, answers for classification of test set. This answer should be tab separated taxonomy text file with header. Evaluation of models is done by accuracy, F1 score, and Matthew Correlation Coefficient (MCC) via sklearn.\n",
    "\n",
    "When running cells, new files will be saved in \"RDPfiles\" directory at the same folder of this jupyter notebook is located. Directory name can be changed. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1. Setting Variables\n",
    "\n",
    "This cell includes necessary packages and variables for preprocessing, training, classifying and scoring. These packages and variables are used in main sciprt.\n",
    "\n",
    "When running cells, new files will be saved in \"RDPfiles\" directory at the same folder of this jupyter notebook is located. Directory name can be changed. \n",
    "3 datasets are required and each should be in csv format with header.\n",
    "Specific path to RDP Classifier is required.\n",
    "Also prediction level can be selected. Training and classification is done with whole taxons, from kingdom to species, but specific level can be choosed in evaluation. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 318,
   "metadata": {},
   "outputs": [],
   "source": [
    "#required packages and variables\n",
    "\n",
    "import os \n",
    "import sys\n",
    "import pandas as pd\n",
    "import string\n",
    "from sklearn.metrics import f1_score, matthews_corrcoef, accuracy_score\n",
    "import numpy as np\n",
    "import time\n",
    "\n",
    "global RDPfiles\n",
    "\n",
    "#variables\n",
    "RDPfiles = \"RDPfiles\"\n",
    "raw_train = \"df_train_0.csv\"\n",
    "raw_val = \"df_val_0.csv\"\n",
    "test = \"df_test_0.csv\"\n",
    "classifier_loc = \"/Users/inkyunpark/vscjava/rdptools/classifier.jar\"\n",
    "confidence_score = 0.8\n",
    "level = 'genus' #ranks = ['kingdom', 'phylum', 'class', 'order', 'family', 'genus', 'species']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Necessary Definition\n",
    "\n",
    "This cell includes three definitions: lineage2taxTrain, addFullLineage, and RDPoutput2score. These definition helps function of main script.\n",
    "\n",
    "Lineage2taxTrain is converting tab separated taxonomy text files into ready4train_taxonomy.txt file. This text file contains the hierarchical taxonomy information in the following format: tax ID * taxon name * parent taxid * depth * rank. \n",
    "- Tax ID is index of specific rank in taxonomy file. If genus *Bacillus* is assigned to 31, there will be no other tax ID with genus *Bacillus*.\n",
    "- Taxon name is a name for each taxonoimc ranks. \n",
    "- Parent taxid is tax ID of higher rank. For example, *Bacillus cereus* is one of *Bacillus* species. If *Bacillus* has tax ID of 43, parent taxid of *Bacillus* becomes 31 too.\n",
    "- Depth means depth of ranks. Depth 0 is always root. Highest ranks, kingdom as an example, has depth of 1. Depth is increased as ranks becomes lower. \n",
    "- Rank is taxonomic ranks.\n",
    "    For example, label for *Bacillus cereus* is 43\\*Bacillus cereus\\*31\\*7\\*Species.\n",
    "\n",
    "AddFullLineage definition makes ready4train_seqs.fasta file. It has similar structure with sequence fasta file, but semicolon separated taxonomy is added to the next of sequence ID. The taxonomy begins with 'Root' and kingdom to lowest taxons. \n",
    "\n",
    "RDPoutput2score definition evaluates accuracy, F1 score and MCC using output.txt and test_taxonomy.txt. Output.txt can be earned fomr classifying test_sequences.txt with trained model. Score is made by comparing output.txt and test_taxonomy.txt, which are prediction and answers respectively."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 319,
   "metadata": {},
   "outputs": [],
   "source": [
    "#NECESSARY DEFINITIONS FOR RDP TRAINING, CLASSIFICATION AND SCORING\n",
    "\n",
    "#codes to make raw sequence and raw taxonomy into ready4rdp input files.\n",
    "#scripts are from https://github.com/GLBRC-TeamMicrobiome/python_scripts. Raw scripts are revised to be used in this concatenated jupyter notebook file.\n",
    "def lineage2taxTrain(raw_taxons):\n",
    "    taxons_list = raw_taxons.strip().split('\\n')\n",
    "    header = taxons_list[0].split('\\t')[1:]#headers = list of ranks\n",
    "    hash = {}#taxon name-id map\n",
    "    ranks = {}#column number-rank map\n",
    "    lineages = []#list of unique lineages\n",
    "\n",
    "    with open(\"{}/ready4train_taxonomy.txt\".format(RDPfiles), \"w\") as f:\n",
    "        hash = {\"Root\":0}#initiate root rank taxon id map\n",
    "        for i in range(len(header)):\n",
    "            name = header[i]\n",
    "            ranks[i] = name\n",
    "        root = ['0', 'Root', '-1', '0', 'rootrank']#root rank info\n",
    "        f.write(\"*\".join(root) +  '\\n')\n",
    "        ID = 0 #taxon id\n",
    "        for line in taxons_list[1:]:\n",
    "            cols = line.strip().split('\\t')[1:]\n",
    "            for i in range(len(cols)):#iterate each column\n",
    "                name = []\n",
    "                for node in cols[:i + 1]:\n",
    "                    node = node.strip()\n",
    "                    if not node in ('-', ''):\n",
    "                        name.append(node)\n",
    "                pName = \";\".join(name[:-1])\n",
    "                if not name in lineages:\n",
    "                    lineages.append(name)\n",
    "                depth = len(name)\n",
    "                name = ';'.join(name)\n",
    "                if name in hash.keys():#already seen this lineage\n",
    "                    continue\n",
    "                try:\n",
    "                    rank = ranks[i]\n",
    "                except KeyError:\n",
    "                    print (cols)\n",
    "                    sys.exit()\n",
    "                if i == 0:\n",
    "                    pName = 'Root'\n",
    "                pID = hash[pName]#parent taxid\n",
    "                ID += 1\n",
    "                hash[name] = ID #add name-id to the map\n",
    "                out = ['%s'%ID, name.split(';')[-1], '%s'%pID, '%s'%depth, rank]\n",
    "                f.write(\"*\".join(out) + '\\n')\n",
    "    f.close()\n",
    "\n",
    "def addFullLineage(raw_taxons, raw_seqs):\n",
    "    hash = {} #lineage map\n",
    "    taxonomy_list = raw_taxons.strip().split('\\n')\n",
    "    for line in taxonomy_list[1:]:\n",
    "        line = line.strip()\n",
    "        cols = line.strip().split('\\t')\n",
    "        lineage = ['Root']\n",
    "        for node in cols[1:]:\n",
    "            node = node.strip()\n",
    "            if not (node == '-' or node == ''):\n",
    "                lineage.append(node)\n",
    "        ID = cols[0]\n",
    "        lineage = ';'.join(lineage).strip()\n",
    "        hash[ID] = lineage\n",
    "    sequence_list = raw_seqs.strip().split('\\n')\n",
    "    with open(\"{}/ready4train_seqs.fasta\".format(RDPfiles), \"w\") as f:\n",
    "        for line in sequence_list:\n",
    "            line = line.strip()\n",
    "            if line == '':\n",
    "                continue\n",
    "            if line[0] == '>':\n",
    "                ID = line.strip().split()[0].replace('>', '')\n",
    "                lineage = hash[ID]\n",
    "                f.write('>' + ID + '\\t' + lineage + '\\n')\n",
    "            else:\n",
    "                f.write(line.strip() + '\\n')\n",
    "    f.close()\n",
    "\n",
    "#codes to make score from prediction and true values.\n",
    "#sklearn.metrics tools are used to make calcuation.\n",
    "def RDPoutput2score(pred_file, true_file, level, cf):\n",
    "    taxon_list = []\n",
    "    ranks = ['kingdom', 'phylum', 'class', 'order', 'family', 'genus', 'species']\n",
    "    level = ranks.index(level)\n",
    "\n",
    "    pred = pd.read_csv(pred_file, sep=\"\\t\", header=None)\n",
    "    pred.drop(pred.columns[1:level+5+2*level],  axis = 'columns', inplace=True)\n",
    "    pred.drop(pred.columns[4:], axis = 'columns', inplace=True)\n",
    "    \n",
    "    pred_dict = {}\n",
    "    for index, row in pred.iterrows():\n",
    "        row = row.tolist()\n",
    "        if row[1] not in taxon_list:\n",
    "            taxon_list += [row[1]]\n",
    "        if float(row[3]) >= cf:\n",
    "            pred_dict[row[0]] = row[1]\n",
    "\n",
    "    true = pd.read_csv(true_file, sep=\"\\t\", header=None)\n",
    "    true_dict = {}\n",
    "    for index, row in true.iterrows():\n",
    "        true_dict[row[0]] = row[level+1]\n",
    "        if row[level+1] not in taxon_list:\n",
    "            taxon_list += [row[level+1]]\n",
    "\n",
    "\n",
    "    y_pred, y_true = [], []\n",
    "    for i in pred_dict.keys():\n",
    "        y_pred.append(taxon_list.index(pred_dict[i]))\n",
    "        y_true.append(taxon_list.index(true_dict[i]))\n",
    "\n",
    "    acc = accuracy_score(y_true, y_pred)\n",
    "    f1 = f1_score(y_true, y_pred, average='weighted') # works well if both y_true_decode and y_pred are list.\n",
    "    mcc = matthews_corrcoef(y_true, y_pred)\n",
    "\n",
    "    print(\"accuracy is {}\\nf1 score is {}\\nMCC score is {}\".format(acc, f1, mcc))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Main\n",
    "\n",
    "Below cells are main script.\n",
    "\n",
    "In this cell, first, new directory is made to save required materials here. Then preprocessing steps are proceed. \n",
    "1. Call three datasets and merges train and validation sets into one train.\n",
    "2. Create raw_seqs and raw_taxons. These will be utilized in lineage2taxTrain and addFullLineage to become ready4train files, which are input for training new RDP models.\n",
    "3. test_sequcnes.fasta and test_taxonomy.txt are made from test dataset. These are used in evaluation of new trained models.\n",
    "4. lineage2taxTrain and addFullLineage to make ready4train files.\n",
    "\n",
    "End of preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 320,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "data preprocessing for RDP is done\n"
     ]
    }
   ],
   "source": [
    "#main_RDP.py\n",
    "\n",
    "os.system(\"mkdir {}\".format(RDPfiles))\n",
    "\n",
    "#merge train and validation database into one.\n",
    "train = pd.concat(map(pd.read_csv, [raw_train, raw_val]))\n",
    "test = pd.read_csv(test)\n",
    "\n",
    "#change train and test dataframe into tab separated taxonomy and sequence string\n",
    "#taxnomy file is tab separated string\n",
    "#sequence file is fasta format with sequence ID and sequence\n",
    "raw_seqs = ''\n",
    "raw_taxons = 'SeqId\tKingdom\tPhylum\tClass\tOrder\tFamily\tGenus\tSpecies' + '\\n'\n",
    "for index, row in train.iterrows():\n",
    "    taxons = row.tolist()\n",
    "    raw_seqs += '>' + taxons[0] + '\\n' + taxons[-1] + '\\n'\n",
    "    raw_taxons += '\\t'.join(taxons[:-1]) + '\\n'\n",
    "\n",
    "#change test dataframe to test sequence into text and fasta files respectively to be utilized by RDP\n",
    "#taxnomy file is tab separated and saved in text file\n",
    "#sequence file is fasta format with sequence ID and sequence and saved in fasta file\n",
    "with open(\"{}/test_sequences.fasta\".format(RDPfiles), \"w\") as seq_f, open(\"{}/test_taxonomy.txt\".format(RDPfiles), \"w\") as tax_f:\n",
    "    for index,row in test.iterrows():\n",
    "        taxons = row.tolist()\n",
    "        seq_f.write('>' + taxons[0] + '\\n' + taxons[-1] + '\\n')\n",
    "        tax_f.write('\\t'.join(taxons[:-1]) + '\\n')\n",
    "    seq_f.close()\n",
    "    tax_f.close()\n",
    "\n",
    "#change raw taxonomy and sequence files to ready4rdp trainable files\n",
    "lineage2taxTrain(raw_taxons)\n",
    "addFullLineage(raw_taxons, raw_seqs)\n",
    "\n",
    "print(\"data preprocessing for RDP is done\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here, new RDP classifier is trained and evaluated with test sets.\n",
    "\n",
    "Training is done with use of train function of RDP Classifier. New models is saved in training_files directory with four files of weights. New file, rRNAClassifier.properties, is needed to be used as a bridge between these four files and RDP Classifier.\n",
    "Classification is done with classify function of RDP Classifier. Option -o leads RDP Classifier to use new training models. Prediction will be written on output.txt.\n",
    "Time for training and classification are meausred.\n",
    "\n",
    "Using RDPoutput2score, accuracy, F1 score and MCC are made. These values will be used to compare with other deep learning models. Users can choose specific level of ranks to evaluate prediction."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 321,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RDP training time is 23.329681158065796 seconds\n",
      "RDP classification time is 53.412200927734375 seconds\n",
      "accuracy is 0.9754299754299754\n",
      "f1 score is 0.9756451668945949\n",
      "MCC score is 0.9752839861577162\n"
     ]
    }
   ],
   "source": [
    "#train new RDP models and add necessary file.\n",
    "t0 = time.time()\n",
    "os.system(\"java -Xmx10g -jar {} train -o {}/training_files -s {}/ready4train_seqs.fasta -t {}/ready4train_taxonomy.txt\".format(classifier_loc, RDPfiles, RDPfiles, RDPfiles))\n",
    "with open(\"{}/training_files/rRNAClassifier.properties\".format(RDPfiles), \"w\") as f:\n",
    "    f.write(\"bergeyTree=bergeyTrainingTree.xml\" + '\\n' + \"probabilityList=genus_wordConditionalProbList.txt\" + '\\n' + \"probabilityIndex=wordConditionalProbIndexArr.txt\" + '\\n' + \"wordPrior=logWordPrior.txt\" + '\\n' + \"classifierVersion=RDP Naive Bayesian rRNA Classifier Version 2.5, May 2012 \")\n",
    "    f.close()\n",
    "t1 = time.time()\n",
    "print(\"RDP training time is {} seconds\".format(t1-t0))\n",
    "\n",
    "#classify test sequences\n",
    "t0 = time.time()\n",
    "os.system(\"java -Xmx10g -jar {} classify -t {}/training_files/rRNAClassifier.properties  -o {}/output.txt {}/test_sequences.fasta\".format(classifier_loc, RDPfiles, RDPfiles, RDPfiles))\n",
    "t1 = time.time()\n",
    "print(\"RDP classification time is {} seconds\".format(t1-t0))\n",
    "\n",
    "#scoring\n",
    "RDPoutput2score(\"{}/output.txt\".format(RDPfiles), \"{}/test_taxonomy.txt\".format(RDPfiles), level, confidence_score)"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "2f75e4b1df7d8e8732b06c42017abb45e73d9d85f7100dfbe870bbbf5194a98c"
  },
  "kernelspec": {
   "display_name": "Python 3.10.3 64-bit ('css-3.10.3')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.3"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
