{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 300,
   "metadata": {},
   "outputs": [],
   "source": [
    "#required packages and variables\n",
    "\n",
    "import os \n",
    "import sys\n",
    "import pandas as pd\n",
    "import string\n",
    "from sklearn.metrics import f1_score, matthews_corrcoef, accuracy_score\n",
    "import numpy as np\n",
    "\n",
    "#variables\n",
    "raw_train = \"df_train_0.csv\"\n",
    "raw_val = \"df_val_0.csv\"\n",
    "test = \"df_test_0.csv\"\n",
    "classifier_loc = \"/Users/inkyunpark/vscjava/rdptools/classifier.jar\"\n",
    "training_files_loc = \"training_files\"\n",
    "classified_loc = \"output.txt\"\n",
    "confidence_score = 0.8\n",
    "level = 'genus' #ranks = ['kingdom', 'phylum', 'class', 'order', 'family', 'genus', 'species']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 301,
   "metadata": {},
   "outputs": [],
   "source": [
    "#NECESSARY DEFINITIONS FOR RDP TRAINING, CLASSIFICATION AND SCORING\n",
    "\n",
    "#codes to make raw sequence and raw taxonomy into ready4rdp input files.\n",
    "#scripts are from https://github.com/GLBRC-TeamMicrobiome/python_scripts and revised to be used in this concatenated jupyter notebook file.\n",
    "\n",
    "def lineage2taxTrain(raw_taxons):\n",
    "    taxons_list = raw_taxons.strip().split('\\n')\n",
    "    header = taxons_list[0].split('\\t')[1:]#headers = list of ranks\n",
    "    hash = {}#taxon name-id map\n",
    "    ranks = {}#column number-rank map\n",
    "    lineages = []#list of unique lineages\n",
    "\n",
    "    with open(\"RDP_files/ready4train_taxonomy.txt\", \"w\") as f:\n",
    "        hash = {\"Root\":0}#initiate root rank taxon id map\n",
    "        for i in range(len(header)):\n",
    "            name = header[i]\n",
    "            ranks[i] = name\n",
    "        root = ['0', 'Root', '-1', '0', 'rootrank']#root rank info\n",
    "        f.write(\"*\".join(root) +  '\\n')\n",
    "        ID = 0 #taxon id\n",
    "        for line in taxons_list[1:]:\n",
    "            cols = line.strip().split('\\t')[1:]\n",
    "            for i in range(len(cols)):#iterate each column\n",
    "                name = []\n",
    "                for node in cols[:i + 1]:\n",
    "                    node = node.strip()\n",
    "                    if not node in ('-', ''):\n",
    "                        name.append(node)\n",
    "                pName = \";\".join(name[:-1])\n",
    "                if not name in lineages:\n",
    "                    lineages.append(name)\n",
    "                depth = len(name)\n",
    "                name = ';'.join(name)\n",
    "                if name in hash.keys():#already seen this lineage\n",
    "                    continue\n",
    "                try:\n",
    "                    rank = ranks[i]\n",
    "                except KeyError:\n",
    "                    print (cols)\n",
    "                    sys.exit()\n",
    "                if i == 0:\n",
    "                    pName = 'Root'\n",
    "                pID = hash[pName]#parent taxid\n",
    "                ID += 1\n",
    "                hash[name] = ID #add name-id to the map\n",
    "                out = ['%s'%ID, name.split(';')[-1], '%s'%pID, '%s'%depth, rank]\n",
    "                f.write(\"*\".join(out) + '\\n')\n",
    "    f.close()\n",
    "\n",
    "def addFullLineage(raw_taxons, raw_seqs):\n",
    "    hash = {} #lineage map\n",
    "    taxonomy_list = raw_taxons.strip().split('\\n')\n",
    "    for line in taxonomy_list[1:]:\n",
    "        line = line.strip()\n",
    "        cols = line.strip().split('\\t')\n",
    "        lineage = ['Root']\n",
    "        for node in cols[1:]:\n",
    "            node = node.strip()\n",
    "            if not (node == '-' or node == ''):\n",
    "                lineage.append(node)\n",
    "        ID = cols[0]\n",
    "        lineage = ';'.join(lineage).strip()\n",
    "        hash[ID] = lineage\n",
    "    sequence_list = raw_seqs.strip().split('\\n')\n",
    "    with open(\"RDP_files/ready4train_seqs.fasta\", \"w\") as f:\n",
    "        for line in sequence_list:\n",
    "            line = line.strip()\n",
    "            if line == '':\n",
    "                continue\n",
    "            if line[0] == '>':\n",
    "                ID = line.strip().split()[0].replace('>', '')\n",
    "                lineage = hash[ID]\n",
    "                f.write('>' + ID + '\\t' + lineage + '\\n')\n",
    "            else:\n",
    "                f.write(line.strip() + '\\n')\n",
    "    f.close()\n",
    "\n",
    "def RDPoutput2score(pred_file, true_file, level, cf):\n",
    "    taxon_list = ['unclassified']\n",
    "    ranks = ['kingdom', 'phylum', 'class', 'order', 'family', 'genus', 'species']\n",
    "    level = ranks.index(level)\n",
    "\n",
    "    pred_file_loc = 'RDP_files/' + pred_file\n",
    "    pred = pd.read_csv(pred_file_loc, sep=\"\\t\", header=None)\n",
    "    pred.drop(pred.columns[1:level+5+2*level],  axis = 'columns', inplace=True)\n",
    "    pred.drop(pred.columns[4:], axis = 'columns', inplace=True)\n",
    "    \n",
    "    pred_dict = {}\n",
    "    for index, row in pred.iterrows():\n",
    "        row = row.tolist()\n",
    "        if row[1] not in taxon_list:\n",
    "            taxon_list += [row[1]]\n",
    "        if float(row[3]) >= cf:\n",
    "            pred_dict[row[0]] = row[1]\n",
    "\n",
    "    true = pd.read_csv(true_file, sep=\"\\t\", header=None)\n",
    "    true_dict = {}\n",
    "    for index, row in true.iterrows():\n",
    "        true_dict[row[0]] = row[level+1]\n",
    "        if row[level+1] not in taxon_list:\n",
    "            taxon_list += [row[level+1]]\n",
    "\n",
    "\n",
    "    y_pred, y_true = [], []\n",
    "    for i in pred_dict.keys():\n",
    "        y_pred.append(taxon_list.index(pred_dict[i]))\n",
    "        y_true.append(taxon_list.index(true_dict[i]))\n",
    "\n",
    "    acc = accuracy_score(y_true, y_pred)\n",
    "    f1 = f1_score(y_true, y_pred, average='weighted') # works well if both y_true_decode and y_pred are list.\n",
    "    mcc = matthews_corrcoef(y_true, y_pred)\n",
    "\n",
    "    print(\"accuracy is {}\\nf1 score is {}\\nMCC score is {}\".format(acc, f1, mcc))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 302,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mkdir: RDP_files: File exists\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "accuracy is 0.9754299754299754\n",
      "f1 score is 0.9756451668945949\n",
      "MCC score is 0.9752839861577162\n"
     ]
    }
   ],
   "source": [
    "#main_RDP.py\n",
    "\n",
    "os.system(\"mkdir RDP_files\")\n",
    "\n",
    "#merge train and validation database into one.\n",
    "train = pd.concat(map(pd.read_csv, [raw_train, raw_val]))\n",
    "test = pd.read_csv(test)\n",
    "\n",
    "#change train and test dataframe into taxonomy and sequence strings\n",
    "raw_seqs = ''\n",
    "raw_taxons = 'SeqId\tKingdom\tPhylum\tClass\tOrder\tFamily\tGenus\tSpecies' + '\\n'\n",
    "for index, row in train.iterrows():\n",
    "    taxons = row.tolist()\n",
    "    raw_seqs += '>' + taxons[0] + '\\n' + taxons[-1] + '\\n'\n",
    "    raw_taxons += '\\t'.join(taxons[:-1]) + '\\n'\n",
    "\n",
    "#change raw taxonomy and sequence files to ready4rdp trainable files\n",
    "lineage2taxTrain(raw_taxons)\n",
    "addFullLineage(raw_taxons, raw_seqs)\n",
    "\n",
    "#change test dataframe to test sequence files to be classified by RDP\n",
    "with open(\"RDP_files/test_sequences.fasta\", \"w\") as seq_f, open(\"RDP_files/test_taxonomy.txt\", \"w\") as tax_f:\n",
    "    for index,row in test.iterrows():\n",
    "        taxons = row.tolist()\n",
    "        seq_f.write('>' + taxons[0] + '\\n' + taxons[-1] + '\\n')\n",
    "        tax_f.write('\\t'.join(taxons[:-1]) + '\\n')\n",
    "    seq_f.close()\n",
    "    tax_f.close()\n",
    "\n",
    "#train RDP models and add necessary file.\n",
    "os.system(\"java -Xmx10g -jar {} train -o RDP_files/{} -s RDP_files/ready4train_seqs.fasta -t RDP_files/ready4train_taxonomy.txt\".format(classifier_loc, training_files_loc))\n",
    "with open(\"RDP_files/{}/rRNAClassifier.properties\".format(training_files_loc), \"w\") as f:\n",
    "    f.write(\"bergeyTree=bergeyTrainingTree.xml\" + '\\n' + \"probabilityList=genus_wordConditionalProbList.txt\" + '\\n' + \"probabilityIndex=wordConditionalProbIndexArr.txt\" + '\\n' + \"wordPrior=logWordPrior.txt\" + '\\n' + \"classifierVersion=RDP Naive Bayesian rRNA Classifier Version 2.5, May 2012 \")\n",
    "    f.close()\n",
    "\n",
    "#classify test sequences\n",
    "os.system(\"java -Xmx10g -jar {} classify -t RDP_files/{}/rRNAClassifier.properties  -o RDP_files/{} RDP_files/test_sequences.fasta\".format(classifier_loc, training_files_loc, classified_loc))\n",
    "\n",
    "#scoring\n",
    "RDPoutput2score(classified_loc, \"RDP_files/test_taxonomy.txt\", level, confidence_score)"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "2f75e4b1df7d8e8732b06c42017abb45e73d9d85f7100dfbe870bbbf5194a98c"
  },
  "kernelspec": {
   "display_name": "Python 3.10.3 64-bit ('css-3.10.3')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.3"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
